src/envoy/http/authnv2/filter_test.cc

```diff
@@ -0,0 +1,205 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/http/authnv2/filter.h"
+
+#include "common/common/base64.h"
+#include "common/http/header_map_impl.h"
+#include "common/stream_info/stream_info_impl.h"
+#include "extensions/filters/http/well_known_names.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "src/envoy/http/authnv2/test_utils.h"
+#include "src/envoy/utils/authn.h"
+#include "src/envoy/utils/filter_names.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/mocks/ssl/mocks.h"
+#include "test/test_common/simulated_time_system.h"
+#include "test/test_common/test_time.h"
+#include "test/test_common/utility.h"
+
+using testing::_;
+using testing::AtLeast;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::Return;
+using testing::ReturnRef;
+using testing::StrictMock;
+
+namespace iaapi = istio::authentication::v1alpha1;
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+namespace {
+
+// Payload data to inject. Note the iss claim intentionally set different from
+// kJwtIssuer.
+static const std::string kMockJwtPayload = R"EOF(
+{
+    "iss":"https://example.com",
+    "sub":"test@example.com",
+    "exp":2001001001,
+    "aud": "example_service"
+}
+)EOF";
+
+class AuthnV2FilterTest : public testing::Test {
+ public:
+  AuthnV2FilterTest()
+      : request_headers_{{":method", "GET"}, {":path", "/"}},
+        ssl_(std::make_shared<NiceMock<Ssl::MockConnectionInfo>>()),
+        test_time_(),
+        stream_info_(Http::Protocol::Http2, test_time_.timeSystem()) {}
+
+  void initialize(const std::vector<std::string>& jwts) {
+    // Jwt metadata filter input setup.
+    const std::string jwt_name =
+        Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn;
+    auto& metadata = stream_info_.dynamicMetadata();
```

Comment:
Please stop using dynamic metadata for any performance-critical work (e.g. this seems to be used 100% of time).
It has huge performance overhead. Use filter state instead.

Reply:
The purpose of the filter is to feed data into rbac filter.

My understanding is that we'll need to use RBAC filter to `metadata matcher` to enforce the policy, https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/rbac/v2/rbac.proto#envoy-api-msg-config-rbac-v2-rbac

I don't know if the filter state can be shared and supported by rbac filter.

Follow-up question, why the filter state is more performant than metadata?  and what are the rough number?

Reply:
Got it, I think we'll need to extend RBAC matcher to support filter state. We found that protobuf struct would consume up to 10% of CPU time just to pass a value between two places. It's just very inefficient.